File: 0096/recursion_with_memoization.cpp
#include <cstdlib>
#include <vector>

#include "gtest/gtest.h"

using namespace std;

class Solution {
 public:
  int numTrees(int n) {
    if (n == 0) { return 0; }
    vector<int> memo(n+1, -1);
    return recursive(n, memo);
  }

  int recursive(int n, vector<int>& memo) {
    if (n == 0 || n == 1) { return 1; }
    if (memo[n] != -1) { return memo[n]; }
    int ret = 0;
    for (int left = 0; left < n; left++) {
      int right = n - 1 - left;
      ret += recursive(left, memo) * recursive(right, memo);
    }
    return memo[n] = ret;
  }
};

int main() {
  Solution sol;
  int n, expected;

  n = 3, expected = 5;
  EXPECT_EQ(expected, sol.numTrees(n));

  n = 0, expected = 0;
  EXPECT_EQ(expected, sol.numTrees(n));

  n = 1, expected = 1;
  EXPECT_EQ(expected, sol.numTrees(n));

  n = 2, expected = 2;
  EXPECT_EQ(expected, sol.numTrees(n));

  exit(EXIT_SUCCESS);
}

File: 0096/recursion_with_memoization.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn num_trees(n: i32) -> i32 {
        let mut memo: HashMap<i32, i32> = HashMap::new();
        memo.entry(0).or_insert(1);
        memo.entry(1).or_insert(1);
        return Solution::recursion(n, &mut memo);
    }

    fn recursion(n: i32, memo: &mut HashMap<i32, i32>) -> i32 {
        match memo.get(&n) {
            Some(num) => return *num,
            None => {}
        }
        let mut ret = 0;
        for i in 0..n {
            let num_left = Solution::recursion(i, memo);
            let num_right = Solution::recursion(n - i - 1, memo);
            ret += num_left * num_right;
        }
        memo.entry(n).or_insert(ret);
        return ret;
    }
}

fn main() {
    let (n, expected) = (3, 5);
    assert_eq!(Solution::num_trees(n), expected);

    let (n, expected) = (1, 1);
    assert_eq!(Solution::num_trees(n), expected);

    let (n, expected) = (19, 1767263190);
    assert_eq!(Solution::num_trees(n), expected);
}
